---
title: "kSanity-VIRGO metagenomics counts QC"
author: Laura Vermeren & Symul
date: today
format: 
   html:
     page-layout: full
     code-fold: true
     toc: true
     toc-location: left
     toc-depth: 5
     embed-resources: true
execute:
  cache: refresh # true refresh false
  warning: false
knitr:
  opts_chunk:
    out.width: "100%"
editor: source
---



## Load the data


VIRGO and kSanity data combined (counts)

## Create a SummarizedExperiment object

(copy-paste from the other notebook)

(include a total count per sample in the colData `total_non_human_reads`)

assay_name = `counts`


## Exploratory & QC analyses

Total number of counts per sample

(Exclude failed samples)

...

## Build a taxonomic table


## Add relative abundances


assay_name = `rel_ab`


## Additional `SE` with bacteria only counts and relative abundances

Take the counts assay from SE1; subset the feature that are bacteria;
compute `total_bacterial_reads` and `rel_ab`


## Save `SummarizedExperiment` objects

Save the `SE` objects to disk

