---
title: "kSanity-VIRGO metagenomics counts QC"
author: Laura Vermeren & Symul
date: today
format: 
   html:
     page-layout: full
     code-fold: true
     toc: true
     toc-location: left
     toc-depth: 5
     embed-resources: true
execute:
  cache: refresh # true refresh false
  warning: false
knitr:
  opts_chunk:
    out.width: "100%"
editor: source
---


```{r}
#| warning: false

library(tidyverse)
library(magrittr)
library(gt)
library(patchwork)
library(SummarizedExperiment)
library(tidySummarizedExperiment)

# tmp <- fs::dir_map("R scripts mg/", source)
tmp <- fs::dir_map("R scripts/", source)

theme_set(theme_light())
```

## Load the data

```{r}

simulated <- TRUE

if (simulated == TRUE) {
  simulated_data_dir <- get_simulated_data_dir()
  mg_dir <- str_c(simulated_data_dir, "03 metagenomics combined/")
  counts <- read_csv(str_c(mg_dir, "mg_combined.csv"))
  manifest <- read_csv(str_c(mg_dir, "mg_combined_manifest.csv"))
} else {
  VIBRANT_dropbox <- get_VIBRANT_Dropbox_dir()
  mg_dir <- str_c(VIBRANT_dropbox, "13_VIBRANT Metagenomics/")
}

```


## Create a SummarizedExperiment object


```{r}

mg_to_SE <- function(counts, mg_combined, mg_manifest, LBP_strain_info){
  mg_manifest <- 
    mg_manifest |> 
    mutate(sample_id = str_c(pid, "_", visit_code)) 
  
  if (any(duplicated(mg_manifest$sample_id))) stop("Duplicated `uid` in `mg_manifest`")
  
  # mg <- mg |> filter(!is.na(`Lactobacillus crispatus`))
  counts <- counts |> left_join(mg_manifest, by = join_by(barcode))
  tmp <- counts |> dplyr::select(barcode, sample_id) |> distinct()
  if (any(duplicated(tmp$sample_id))) stop("Duplicated `uid` in `mg`")
  
  assay_counts <- 
    counts |> 
    dplyr::select(sample_id, everything()) |> 
    dplyr::select(-c(pid, visit_code, barcode, total_reads)) |> 
    as.data.frame() |> 
    column_to_rownames("sample_id") |> 
    drop_na() |> 
    t() 
  
  ## Checking if total_reads match the sum accross taxa
  tibble(
    sample_id = colnames(assay_counts),
    sum_counts = colSums(assay_counts)
  ) |> 
    left_join(
      counts |> select(sample_id, total_reads),
      by = join_by(sample_id)
    ) |> 
    ggplot() +
    aes(x = sum_counts, y = total_reads) +
    geom_abline(slope = 1, intercept = 0, col = "steelblue1") +
    geom_point(alpha = 0.5, size = 0.3)
    
  assay_prop_of_Lc <- 
    (t(assay_counts)/colSums(assay_counts)) |> 
    t()
 
  se_coldata <-
    tibble(
      sample_id = colnames(assay_counts),
      sum_counts = colSums(assay_counts)
    ) |> 
    left_join(
      counts |> select(sample_id, barcode, total_reads), by = join_by(sample_id)
      ) |> 
    select(barcode, everything()) |> 
    as.data.frame() |> 
    column_to_rownames("sample_id")

  se_rowdata <- 
    mg |> 
    dplyr::select(-c(barcode, pid, visit_code, sample_id)) |> 
    colnames() |> 
    as.data.frame() |> 
    setNames("strain") |> 
    distinct() |> 
    left_join(
      LBP_strain_info,
      by = join_by(strain == strain_id)
    ) 
  
  
  # Harmonization of the order of samples and feature
  
  sorted_sample_ids <- sort(as.character(se_coldata$sample_id))
  assay_prop_of_Lc <- assay_prop_of_Lc[, sorted_sample_ids]
  assay_unique_kmers <- assay_unique_kmers[, sorted_sample_ids]
  se_coldata <- se_coldata[match(sorted_sample_ids, se_coldata$sample_id), ]

  sorted_strains <- assay_prop_of_Lc |> rownames() # sort(as.character(se_rowdata$strain))
  assay_prop_of_Lc <- assay_prop_of_Lc[sorted_strains, ]
  assay_unique_kmers <- assay_unique_kmers[sorted_strains, ]
  se_rowdata <- se_rowdata[match(sorted_strains, se_rowdata$strain), ]

  se_rowdata <- as(se_rowdata, "DataFrame")
  se_coldata <- as(se_coldata, "DataFrame")

  SummarizedExperiment::SummarizedExperiment(
    assays = list(prop_of_Lc = assay_prop_of_Lc, unique_kmers = assay_unique_kmers),
    rowData = se_rowdata,
    colData = se_coldata
  )
}

```


(copy-paste from the other notebook)

(include a total count per sample in the colData `total_non_human_reads`)

assay_name = `counts`


## Exploratory & QC analyses

Total number of counts per sample

(Exclude failed samples)

...

## Build a taxonomic table


## Add relative abundances


assay_name = `rel_ab`


## Additional `SE` with bacteria only counts and relative abundances

Take the counts assay from SE1; subset the feature that are bacteria;
compute `total_bacterial_reads` and `rel_ab`


## Save `SummarizedExperiment` objects

Save the `SE` objects to disk

