---
title: "kSanity-VIRGO metagenomics counts QC"
author: Laura Vermeren & Symul
date: today
format: 
   html:
     page-layout: full
     code-fold: true
     toc: true
     toc-location: left
     toc-depth: 5
     embed-resources: true
execute:
  cache: refresh # true refresh false
  warning: false
knitr:
  opts_chunk:
    out.width: "100%"
editor: source
---


```{r}
#| warning: false

library(tidyverse)
library(magrittr)
library(gt)
library(patchwork)
library(SummarizedExperiment)
library(tidySummarizedExperiment)
library(labelled)
library(vegan)
library(RColorBrewer)

# tmp <- fs::dir_map("R scripts mg/", source)
tmp <- fs::dir_map("R scripts/", source)

theme_set(theme_light())
```

## Load the data

```{r}

data_source <- "real"
data_dir <- get_data_dir(data_source)

if (data_source == TRUE) {
  mg_dir <- str_c(data_dir, "03 metagenomics combined/")
  counts <- read_csv(str_c(mg_dir, "mg_combined.csv"))
  manifest <- read_csv(str_c(mg_dir, "mg_combined_manifest.csv"))
} else {
  mg_dir <- str_c(data_dir, "02 Metagenomics/")
  counts <- read_csv(str_c(mg_dir, "MVIBR_kSanityVIRGO2_ReadCounts_20250404.csv"))
  counts_corr <- read_csv(str_c(mg_dir, "MVIBR_kSanityVIRGO2_taxaGLcor_20250404.csv"))
  relabs <- read_csv(str_c(mg_dir, "MVIBR_kSanityVIRGO2_RelAbund_20250404.csv"))
  technical_metadata <- read_csv(str_c(mg_dir, "VIBRANT_MG_technicalMetaData_20250404.csv"))
  LBP_strain_info <- readxl::read_xlsx(str_c(data_dir, "00 Trial Data/IsolateNumbers.xlsx"))
}

```

The names of the strains do not correspond exactly to those in the metagenomics files. We will modify the names of the strains in *LBP_strain_info* to match them.

```{r}

LBP_strain_info <- 
  LBP_strain_info |> 
  mutate(strain_id = `VMRC ID`, 
         LBP = ifelse(is.na(LC106), "LC-115", "LC-106 & LC-115") |> factor(), 
         strain_origin = `Geographic source` |> factor()
  ) |>
  dplyr::rename(Biose_ID = `Biose ID`) |> #VMRC_ID = `VMRC ID` 
  dplyr::select(strain_id, LBP, strain_origin, Biose_ID) |> 
  arrange(strain_origin, LBP) |> 
  mutate(strain_id = strain_id |> fct_inorder()) |> 
  dplyr::select(strain_id, LBP, strain_origin, contains("ID")) |> 
  mutate(strain_id = sub("_.*", "", strain_id),
         strain_id = case_when(strain_id == "CC0028A1" ~ "C0028A1", 
                               TRUE ~ strain_id
    ))

LBP_strain_info |>   
  set_variable_labels(
    strain_id = "Strain ID", 
    strain_origin = "Strain origin", 
    Biose_ID = "Biose ID", 
    #VMRC_ID = "VMRC ID"
  ) |> 
  gt(caption = "LBP strain information") |> 
  tab_style(style = cell_text(weight = "bold"),
            locations = cells_column_labels())


```

## Create a SummarizedExperiment object


>NOTE : not possible for the moment because the number of columns does not correspond between the different assays. `counts` includes an extra column "multiGenera".
For now, we remove the "MuliGenra" from the `counts` data frame.

```{r}

mg_to_SE <- function(counts, counts_corr, relabs, LBP_strain_info, technical_metadata) {
  
  technical_metadata <- 
    technical_metadata |> 
    filter(is.na(LibrarySequencedTwice)) 
  
  # suppress "multiGenera" from counts
  counts <- 
    counts |> 
    select(-c(MultiGenera))

  assay_counts <- 
    counts |> 
    dplyr::select(-c(CST, subCST, score)) |> 
    as.data.frame() |> 
    column_to_rownames("sampleID") |> 
    drop_na() |> 
    t() 
  
  assay_counts_corr <- 
    counts_corr |> 
    dplyr::select(-c(CST, subCST, score)) |>
    as.data.frame() |>
    column_to_rownames("sampleID") |>
    t()
  
  assay_relative_ab <- 
    relabs |> 
    dplyr::select(-c(CST, subCST, score)) |>
    as.data.frame() |> 
    column_to_rownames("sampleID") |>
    t()
    
  
  # include a total reads per sample in the colData 
  se_coldata <-
    counts |> 
    rowwise() |> 
    mutate(total_non_human_reads = sum(c_across(-c(sampleID, CST, subCST, score)))) |>
    select(sampleID, total_non_human_reads) |> 
    left_join(
      technical_metadata |> filter(is.na(LibrarySequencedTwice)), 
      by = c("sampleID" = "UID")
    ) |>
    as.data.frame()
  
  se_rowdata <- 
    counts |> 
    dplyr::select(-c(sampleID, CST, subCST, score)) |>
    colnames() |> 
    as.data.frame() |> 
    setNames("strain") |> 
    distinct() |> 
    left_join(
      LBP_strain_info,
      by = join_by(strain == strain_id)
    ) 
  
  # Harmonization of the order of samples and feature
  
  sorted_sample_ids <- sort(as.character(se_coldata$sampleID))
  assay_counts <- assay_counts[, sorted_sample_ids]
  assay_counts_corr <- assay_counts_corr[, sorted_sample_ids]
  assay_relative_ab <- assay_relative_ab[, sorted_sample_ids]
  se_coldata <- se_coldata[match(sorted_sample_ids, se_coldata$sampleID), ]

  sorted_strains <- assay_relative_ab |> rownames() # sort(as.character(se_rowdata$strain))
  assay_counts <- assay_counts[sorted_strains, ]
  assay_counts_corr <- assay_counts_corr[sorted_strains, ]
  assay_relative_ab <- assay_relative_ab[sorted_strains, ]
  se_rowdata <- se_rowdata[match(sorted_strains, se_rowdata$strain), ]

  SummarizedExperiment::SummarizedExperiment(
    assays = list(counts = assay_counts, counts_corr = assay_counts_corr, relabs = assay_relative_ab),
    rowData = se_rowdata,
    colData = se_coldata
  )
}

```

```{r}

SE_mg <- mg_to_SE(counts, counts_corr, relabs, LBP_strain_info, technical_metadata)

```


(include a total count per sample in the colData `total_non_human_reads`) > OK

assay_name = `counts`


## Exploratory & QC analyses

```{r}
SE_mg |> as.tibble()
```


###Total number of counts/relative abundances per sample

**Total number of counts and corrected counts per sample**

```{r}

SE_mg |> 
  as.tibble() |> 
  group_by(.sample) |> 
  summarise(total_count = sum(counts)) |>
  ggplot(aes(x = total_count)) +
  geom_histogram() +
  scale_x_log10() +
  labs(x = "Total number of counts per sample", 
       y = "Number of samples")

SE_mg |> 
  as.tibble() |> 
  group_by(.sample) |> 
  summarise(total_count = sum(counts_corr)) |>
  ggplot(aes(x = total_count)) +
  geom_histogram() +
  scale_x_log10() +
  labs(x = "Total number of corected counts per sample", 
       y = "Number of samples")

```
A sample has a very small total number of counts

```{r}

SE_mg |> 
  group_by(.sample) |>
  as.tibble() |> 
  filter(total_non_human_reads < 1e5) |>
  distinct(.sample) |> 
  print()

```

**Total proportion per sample**

```{r}

SE_mg |> 
  as.tibble() |> 
  group_by(.sample) |>
  summarise(total_relab = sum(relabs)) |> 
  ggplot(aes(x = total_relab)) +
  geom_histogram() + 
  labs(x = "Total proportion per sample", 
       y = "Number of samples")


```

### Missing values

```{r}

SE_mg |> 
  as.tibble() |> 
  ggplot(aes(x=.feature, y = .sample, fill = is.na(counts))) + 
  geom_tile() + 
  scale_x_discrete("Strains") +
  scale_y_discrete("Species",breaks = NULL) +
  labs(title = "Missing values in readcounts") +
  scale_fill_manual(
    name = "Statut",
    values = c("TRUE" = "pink", "FALSE" = "steelblue2"),  
    labels = c("FALSE" = "Not missing", "TRUE" = "Missing")) + 
  theme(axis.text.x = element_blank(),  
        axis.ticks.x = element_blank(), 
        plot.title = element_text(hjust = 0.5))

SE_mg |> 
  as.tibble() |> 
  ggplot(aes(x=.feature, y = .sample, fill = is.na(counts_corr))) + 
  geom_tile() + 
  scale_x_discrete("Species") +
  scale_y_discrete("Samples",breaks = NULL) +
  labs(title = "Missing values in corrected readcounts") +
  scale_fill_manual(
    name = "Statut",
    values = c("TRUE" = "pink", "FALSE" = "steelblue2"),  
    labels = c("FALSE" = "Not missing", "TRUE" = "Missing")) + 
  theme(axis.text.x = element_blank(),  
        axis.ticks.x = element_blank(), 
        plot.title = element_text(hjust = 0.5))

SE_mg |> 
  as.tibble() |> 
  ggplot(aes(x=.feature, y = .sample, fill = is.na(relabs))) + 
  geom_tile() + 
  scale_x_discrete("Species") +
  scale_y_discrete("Samples",breaks = NULL) +
  labs(title = "Missing values in relative abundance") +
  scale_fill_manual(
    name = "Statut",
    values = c("TRUE" = "pink", "FALSE" = "steelblue2"),  
    labels = c("FALSE" = "Not missing", "TRUE" = "Missing")) + 
  theme(axis.text.x = element_blank(),  
        axis.ticks.x = element_blank(), 
        plot.title = element_text(hjust = 0.5))


```

### Proportion of LBP strains per sample

```{r}

SE_mg |> 
  as.tibble() |>  
  filter(!is.na(Biose_ID)) |> 
  ggplot(aes(x = .feature, y = .sample, fill = relabs)) +
  geom_tile() +
  scale_fill_continuous(low = "white", high = "steelblue2") +
  scale_x_discrete("Strains") +
  scale_y_discrete("Samples",breaks = NULL) +
  facet_grid(. ~ LBP + strain_origin, scales = "free_x", space = "free") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1))


```
**Co-occurrence analysis**

Is there a strain that is only present when the others are not?

>TODO : improve plot (same order of taxa, why there is an X in front of LBP names in y ??)

```{r}

presence_absence <- 
  SE_mg |> 
  as.tibble() |>  
  filter(!is.na(Biose_ID)) |> 
  select(.feature, .sample, relabs) |> 
  pivot_wider(names_from = .feature, values_from = relabs, values_fill = 0) |>
  mutate(across(where(is.numeric), ~ ifelse(. > 0, 1, 0))) |> 
  select(-.sample)

cooccurrence <- presence_absence |>
  summarise(across(everything(), sum)) |>
  pivot_longer(cols = everything(), names_to = "taxon", values_to = "presence_count")

cooccurrence_matrix <- presence_absence |>
  summarise(across(everything(), list)) |>
  pivot_longer(cols = everything(), names_to = "taxon", values_to = "presence_list") |>
  mutate(presence_list = map(presence_list, unlist)) |>
  pull(presence_list)

cooccur_mat = matrix(0, nrow = length(cooccurrence_matrix), ncol = length(cooccurrence_matrix))
rownames(cooccur_mat) = colnames(presence_absence)
colnames(cooccur_mat) = colnames(presence_absence)

for (i in 1:length(cooccurrence_matrix)){
  for (j in 1:length(cooccurrence_matrix)){
    cooccur_mat[i,j] = sum(cooccurrence_matrix[[i]] & cooccurrence_matrix[[j]])
  }
}

exclusive_strains <- data.frame(cooccur_mat) |>
  mutate(strain = rownames(cooccur_mat)) |>
  pivot_longer(!strain, names_to = "other_strain", values_to = "cooccur") |>
  group_by(strain) |>
  summarise(total_cooccur = sum(cooccur)) |>
  filter(total_cooccur == 0)

print(exclusive_strains)

cooccur_melted = data.frame(cooccur_mat) |>
  mutate(strain = rownames(cooccur_mat)) |>
  pivot_longer(!strain, names_to = "other_strain", values_to = "cooccur")

ggplot(cooccur_melted, aes(x = strain, y = other_strain, fill = cooccur)) +
  geom_tile() +
  scale_fill_continuous(low = "white", high = "steelblue2") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1))


```
### Control sample

**LBP abundance in control samples**

```{r}

SE_mg |> 
  as.tibble() |>  
  filter(!is.na(Biose_ID))|>
  filter(SampleType != "ClinicalSample") |> 
  ggplot(aes(x = .feature, y = .sample, fill = relabs)) +
  geom_tile() +
  scale_fill_continuous(low = "white", high = "steelblue2") +
  scale_x_discrete("Strains") +
  scale_y_discrete("Samples") +
  facet_grid(SampleType ~ LBP + strain_origin, scales = "free", space = "free") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1))

```

```{r}
#| fig-height: 8

top20_species_all <-
  SE_mg |>
  as.tibble() |>
  group_by(.feature) |>
  summarise(total_relabs = sum(relabs)) |>
  arrange(desc(total_relabs)) |>
  slice_head(n = 20) |>
  group_by(color = case_when(
    .feature %in% LBP_strain_info$strain_id ~ colorRampPalette(c("lightpink", "hotpink3"))(n()),
    str_detect(.feature, "Lactobacillus") ~ colorRampPalette(c("chocolate1","chocolate4"))(n()),
    str_detect(.feature, "Gardnerella") ~ colorRampPalette(c("darkolivegreen1", "darkolivegreen4"))(n()),
    str_detect(.feature, "Prevotella") ~ colorRampPalette(c("lightblue", "deepskyblue3"))(n()),
    TRUE ~ colorRampPalette(c("antiquewhite", "antiquewhite3"))(n())
  )) 

top20_species_control <-
  SE_mg |>
  as.tibble() |>
  filter(SampleType != "ClinicalSample") |>
  group_by(.feature) |>
  summarise(total_relabs = sum(relabs)) |>
  arrange(desc(total_relabs)) |>
  slice_head(n = 20) |>
  group_by(color = case_when(
    .feature %in% LBP_strain_info$strain_id ~ colorRampPalette(c("lightpink", "hotpink3"))(n()),
    str_detect(.feature, "Lactobacillus") ~ colorRampPalette(c("chocolate1","chocolate4"))(n()),
    str_detect(.feature, "Gardnerella") ~ colorRampPalette(c("darkolivegreen1", "darkolivegreen4"))(n()),
    str_detect(.feature, "Prevotella") ~ colorRampPalette(c("lightblue", "deepskyblue3"))(n()),
    TRUE ~ colorRampPalette(c("antiquewhite", "antiquewhite3"))(n())
  )) 


SE_mg |> 
  as.tibble() |>
  filter(.feature %in% all_of(top20_species_all$.feature)) |>
  filter(SampleType != "ClinicalSample") |>
  ggplot(aes(x = .feature, y = .sample, fill = relabs)) +
  geom_tile() +
  scale_fill_continuous(low = "white", high = "steelblue2") +
  scale_x_discrete("Species") +
  scale_y_discrete("Samples") +
  facet_grid(SampleType ~ ., scales = "free", space = "free") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1))

SE_mg |>
  as.tibble() |>
  filter(.feature %in% all_of(top20_species_all$.feature)) |>
  left_join(
    top20_species_all |> select(.feature, color),
    by = ".feature"
  ) |>
  filter(SampleType != "ClinicalSample") |>
  mutate(.feature = factor(.feature, levels = sort(unique(.feature)))) |> 
  ggplot(aes(x = .sample, y = relabs, fill = .feature)) +
  geom_bar(stat = "identity") +
  facet_wrap(~ SampleType, scales = "free_x") +
  labs(
    x = "Samples",
    y = "Relative Abundance",
    title = "Controle sample : Top 20 species of all samples",
    fill = "Species"
  ) +
  scale_x_discrete("Samples", breaks = NULL) +
  scale_fill_manual(
    values = setNames(top20_species_all$color, top20_species_all$.feature),
    breaks = sort(unique(top20_species_all$.feature)), 
    labels = sort(unique(top20_species_all$.feature))  
  ) +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1), 
        plot.title = element_text(hjust = 0.5))

SE_mg |>
  as.tibble() |>
  filter(.feature %in% all_of(top20_species_control$.feature)) |>
  left_join(
    top20_species_control |> select(.feature, color),
    by = ".feature"
  ) |>
  filter(SampleType != "ClinicalSample") |>
  mutate(.feature = factor(.feature, levels = sort(unique(.feature)))) |> 
  ggplot(aes(x = .sample, y = relabs, fill = .feature)) +
  geom_bar(stat = "identity") +
  facet_wrap(~ SampleType, scales = "free_x") +
  labs(
    x = "Samples",
    y = "Relative Abundance",
    title = "Controle sample : Top 20 species of control samples",
    fill = "Species"
  ) +
  scale_x_discrete("Samples", breaks = NULL) +
  scale_fill_manual(
    values = setNames(top20_species_control$color, top20_species_control$.feature),
    breaks = sort(unique(top20_species_control$.feature)), 
    labels = sort(unique(top20_species_control$.feature))  
  ) +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1), 
      plot.title = element_text(hjust = 0.5))



```

### Technical metadata

```{r}

data_plot <-
  technical_metadata |>
  select(-c(UID, Sample, SampleType, Library, Notes, FragmentSize, IndexSet, `Index 1`, `Index 2`, Ext_Lib_Position)) |>
  mutate(across(everything(), as.factor)) |>
  rename(LibraryPool = `Library Pool`,
         Selected4re_extraction = `Selected4re-extraction`)

generate_plot <- function(df, x_var, fill_var) {
  n_levels_fill <- nlevels(df[[fill_var]])
  if (n_levels_fill <= 12) {
    ggplot(df, aes_string(x = x_var, fill = fill_var)) +
      geom_bar(position = "stack") +
      theme_bw() +
      scale_fill_manual(values = brewer.pal(n = n_levels_fill, name = "Set3")) +
      labs(x = x_var, fill = fill_var) +
      theme(axis.text.x = element_text(angle = 90, hjust = 1))
  } else {
    message(cat("Trop de niveaux : x = ",x_var, "fill =", fill_var))
    invisible(NULL) # Retourne NULL pour que walk() ne tente pas d'imprimer un message
  }
}

factor_vars <- names(data_plot)

combinations <- cross(list(x = factor_vars, fill = factor_vars)) %>%
  keep(~ which(.x$x == factor_vars) < which(.x$fill == factor_vars)) 

# Appliquer la fonction de génération de graphique à chaque combinaison
walk(combinations, ~ print(generate_plot(data_plot, .x$x, .x$fill)))

```



### PCoA on counts

```{r}

dist_matrix <-
  vegdist(
    assay(SE_mg, "counts") |> t(), 
    method = "bray")

pcoa <- wcmdscale(dist_matrix, eig = TRUE)
print(pcoa)

pcoa$eig |>
  as.data.frame() |>
  rownames_to_column("axis") |>
  mutate(axis = str_remove(axis, "Eigenvalue"), 
         var_explained = 100*pcoa$eig/sum(pcoa$eig)) |>
  slice(1:10) |>
  gt() |>
  tab_header(title = "Eigenvalues of the PCoA") |>
  cols_label(axis = "Axis", `pcoa$eig` = "Eigenvalue", var_explained = "Variance explained") |>
  fmt_number(columns = vars(`pcoa$eig`, var_explained), decimals = 3)

pcoa$eig |> 
  as_data_frame() |>
  rownames_to_column("axis") |>
  mutate(axis = axis |> as.numeric()) |>
  ggplot(aes(x = axis, y = value)) +
  geom_col() +
  labs(
    x= "Axis",
    y = "Eigenvalue"
  ) + 
  theme_bw()

pcoa_data <- 
  as.data.frame(pcoa$points) |> 
  mutate(sampleID = counts$sampleID) |> 
  left_join(
    technical_metadata |> filter(is.na(LibrarySequencedTwice)), 
    by = c("sampleID" = "UID")
  )

```

```{r}

pcoa_plot <- function(pcoa, color_var, axes = 1:2){
  
  pcoa_data <- pcoa_data |> 
    mutate(!!sym(color_var) := as.factor(!!sym(color_var)))

  pcoa_data |>
    ggplot(aes(x = !!sym(paste0("Dim", axes[1])), y = !!sym(paste0("Dim", axes[2])), color = !!sym(color_var))) +
    geom_point(size = 2, alpha = 0.7) +
    coord_fixed() +
    labs(
      x = paste0("PCoA ", axes[1], " (", round(100 * pcoa$eig[axes[1]] / sum(pcoa$eig), 1), "%)"),
      y = paste0("PCoA ", axes[2], " (", round(100 * pcoa$eig[axes[2]] / sum(pcoa$eig), 1), "%)"),
      color = color_var
    ) + 
    theme_bw()
}


```

#### Type of sample

```{r}

pcoa_plot(pcoa, "SampleType", c(1, 2))
pcoa_plot(pcoa, "SampleType", c(3, 4))
pcoa_plot(pcoa, "SampleType", c(5, 6))

```

```{r}

manova <- adonis2(
  dist_matrix ~ SampleType,
  data = pcoa_data
)

manova

```


#### Plate

```{r}

pcoa_plot(pcoa, "Ext_Lib_Plate", c(1, 2))
pcoa_plot(pcoa, "Ext_Lib_Plate", c(3, 4))
pcoa_plot(pcoa, "Ext_Lib_Plate", c(5, 6))

```

Permutation test 

```{r}

manova <- adonis2(
  dist_matrix ~ Ext_Lib_Plate,
  data = pcoa_data
)

manova
```

#### Batch 

```{r}

pcoa_plot(pcoa, "BioinformaticsProcessingBatch", c(1, 2))
pcoa_plot(pcoa, "BioinformaticsProcessingBatch", c(3, 4))
pcoa_plot(pcoa, "BioinformaticsProcessingBatch", c(5, 6))

```
```{r}

manova <- adonis2(
  dist_matrix ~ BioinformaticsProcessingBatch,
  data = pcoa_data
)

manova

```
#### Library Pool

```{r}

pcoa_plot(pcoa, "Library Pool", c(1, 2))
pcoa_plot(pcoa, "Library Pool", c(3, 4))
pcoa_plot(pcoa, "Library Pool", c(5, 6))

```
```{r}

manova <- adonis2(
  dist_matrix ~ `Library Pool`,
  data = pcoa_data
)

manova
```


#### Lane

```{r}

pcoa_plot(pcoa, "Lane", c(1, 2))
pcoa_plot(pcoa, "Lane", c(3, 4))
pcoa_plot(pcoa, "Lane", c(5, 6))


```
```{r}

manova <- adonis2(
  dist_matrix ~ Lane,
  data = pcoa_data
)

manova

```

### PCoA on relative abundances

```{r}

dist_matrix_relab <-
  vegdist(
    relabs |> dplyr::select(-c(CST, subCST, score)) |> column_to_rownames("sampleID"), 
    method = "bray")

pcoa_relab <- wcmdscale(dist_matrix_relab, eig = TRUE)
print(pcoa_relab)

pcoa_relab$eig |>
  as.data.frame() |>
  rownames_to_column("axis") |>
  mutate(axis = str_remove(axis, "Eigenvalue"), 
         var_explained = 100*pcoa$eig/sum(pcoa_relab$eig)) |>
  slice(1:10) |>
  gt() |>
  tab_header(title = "Eigenvalues of the PCoA") |>
  cols_label(axis = "Axis", `pcoa_relab$eig` = "Eigenvalue", var_explained = "Variance explained") |>
  fmt_number(columns = vars(`pcoa_relab$eig`, var_explained), decimals = 3)

pcoa_relab$eig |> 
  as_data_frame() |>
  rownames_to_column("axis") |>
  mutate(axis = axis |> as.numeric()) |>
  ggplot(aes(x = axis, y = value)) +
  geom_col() +
  labs(
    x= "Axis",
    y = "Eigenvalue"
  ) + 
  theme_bw()

pcoa_relab_data <- 
  as.data.frame(pcoa_relab$points) |> 
  mutate(sampleID = counts$sampleID) |> 
  left_join(
    technical_metadata |> filter(is.na(LibrarySequencedTwice)), 
    by = c("sampleID" = "UID")
  )

```

#### Type of sample

```{r}

pcoa_plot(pcoa_relab, "SampleType", c(1, 2))
pcoa_plot(pcoa_relab, "SampleType", c(1, 2))
pcoa_plot(pcoa_relab, "SampleType", c(1, 2))

```


#### Plate

```{r}

pcoa_plot(pcoa_relab, "Ext_Lib_Plate", c(1, 2))
pcoa_plot(pcoa_relab, "Ext_Lib_Plate", c(3, 4))
pcoa_plot(pcoa_relab, "Ext_Lib_Plate", c(5, 6))

```

```{r}

manova <- adonis2(
  dist_matrix_relab ~ Ext_Lib_Plate,
  data = pcoa_relab_data
)

manova

```


#### Batch 

```{r}

pcoa_plot(pcoa_relab, "BioinformaticsProcessingBatch", c(1, 2))
pcoa_plot(pcoa_relab, "BioinformaticsProcessingBatch", c(3, 4))
pcoa_plot(pcoa_relab, "BioinformaticsProcessingBatch", c(5, 6))

```

```{r}

manova <- adonis2(
  dist_matrix_relab ~ BioinformaticsProcessingBatch,
  data = pcoa_relab_data
)

manova


```
#### Library Pool

```{r}

pcoa_plot(pcoa_relab, "Library Pool", c(1, 2))
pcoa_plot(pcoa_relab, "Library Pool", c(3, 4))
pcoa_plot(pcoa_relab, "Library Pool", c(5, 6))


```

```{r}

manova <- adonis2(
  dist_matrix_relab ~ `Library Pool`,
  data = pcoa_relab_data
)

manova

```


#### Lane

```{r}

pcoa_plot(pcoa_relab, "Lane", c(1, 2))
pcoa_plot(pcoa_relab, "Lane", c(3, 4))
pcoa_plot(pcoa_relab, "Lane", c(5, 6))


```

```{r}

manova <- adonis2(
  dist_matrix_relab ~ Lane,
  data = pcoa_relab_data
)

manova

```



(Exclude failed samples)

...

## Build a taxonomic table


## Add relative abundances


assay_name = `rel_ab`


## Additional `SE` with bacteria only counts and relative abundances

Take the counts assay from SE1; subset the feature that are bacteria;
compute `total_bacterial_reads` and `rel_ab`


## Save `SummarizedExperiment` objects

Save the `SE` objects to disk

